packages/lexical/src/LexicalUtils.ts

```diff
@@ -1084,10 +1084,49 @@ export function isSelectAll(
   return key.toLowerCase() === 'a' && controlOrMeta(metaKey, ctrlKey);
 }
 
-export function $selectAll(): void {
+export function $selectAll(selection?: RangeSelection | null): RangeSelection {
   const root = $getRoot();
-  const selection = root.select(0, root.getChildrenSize());
-  $setSelection($normalizeSelection(selection));
+
+  if (selection && $isRangeSelection(selection)) {
+    // Expand the existing RangeSelection
+    const anchor = selection.anchor;
+    const focus = selection.focus;
+    const anchorNode = anchor.getNode();
+    const topParent = anchorNode.getTopLevelElementOrThrow();
+    const rootNode = topParent.getParentOrThrow();
+    let firstNode = rootNode.getFirstDescendant();
+    let lastNode = rootNode.getLastDescendant();
+    let firstType: 'element' | 'text' = 'element';
+    let lastType: 'element' | 'text' = 'element';
+    let lastOffset = 0;
+
+    if ($isTextNode(firstNode)) {
+      firstType = 'text';
+    } else if (!$isElementNode(firstNode) && firstNode !== null) {
+      firstNode = firstNode.getParentOrThrow();
+    }
+
+    if ($isTextNode(lastNode)) {
+      lastType = 'text';
+      lastOffset = lastNode.getTextContentSize();
+    } else if (!$isElementNode(lastNode) && lastNode !== null) {
+      lastNode = lastNode.getParentOrThrow();
+    }
+
+    if (firstNode && lastNode) {
+      anchor.set(firstNode.getKey(), 0, firstType);
+      focus.set(lastNode.getKey(), lastOffset, lastType);
+    }
+
+    // Normalize and set the updated selection
+    $setSelection($normalizeSelection(selection));
```

Comment:
This implementation shouldn't be calling `$setSelection`, if the user passes in a RangeSelection it's possible they do not want it to be also set (e.g. they are passing it to some other API that takes a selection, but do not want to affect the current editor state).

The `$normalizeSelection` is redundant here because that's what the rest of the code is doing. I'm sure most of it is just copied from the previous implementation, which likely predates `$normalizeSelection`. Since we do have `$normalizeSelection` we might as well simplify it all the way. Maybe something like this:

```suggestion
    const anchorNode = anchor.getNode();
    const topParent = anchorNode.getTopLevelElementOrThrow();
    const rootNode = topParent.getParentOrThrow();
    anchor.set(rootNode.getKey(), 0, 'element');
    focus.set(rootNode.getKey(), rootNode.getChildrenSize(), 'element');
    $normalizeSelection(selection);
```